cfep-09.md

```diff
@@ -0,0 +1,303 @@
+in topo order.
+
+This CFEP is taken from @beckermr's work on [conda-forge.github.io](https://github.com/conda-forge/conda-forge.github.io/issues/712) and @@mariusvniekerk's work in [conda-smithy](https://github.com/conda-forge/conda-forge.github.io/issues/712).
+
+## Motivation
+Currently when we move pinnings we move the pin in the ``conda-forge-pinning-feedstock``.
+Then we start a migration using the ``regro-cf-autotick-bot``.
+This order of operations causes some issues, when new feedstocks
+are created, or existing feedstocks rerendered they receive the new pinnings 
+even if their dependencies are not rebuilt with the new pinnings themselves.
+
+## Specification
+
+### The migrations proceed as follows:
+1. We make a PR into the ``migrations`` folder in the ``conda-forge-pinning-feedstock`` with a new yaml file representing the migration.
+2. Once the change is merged, the bot picks it up, builds a migrator object and begins the migration process
+3. A migration PR is issued for a node only if 
+    1. The node depends on the changed pinnings.
+    2. It has no dependencies that depend on the new pinnings and have not been migrated.
+4. Process 3 continues until we determine that the migration is sufficiently complete and the change is applied to the global pinning file via a PR.
+
+### Implementation Details:
+Given a feedstock with the following pinnings
+
+```yaml
+# Existing pins
+a:
+  - 1.5
+b:
+  - 1.2
+someflag:
+  - disabled
+```
+
+There are four classes of pinning we can do in general.
+
+#### 1. Simple Additive Version Migrations
+
+<table>
+<tr>
+<td>
+
+```yaml
+# Existing pins
+a:
+  - 1.5
+b:
+  - 1.2
+someflag:
+  - disabled
+```
+</td>
+<td> with </td>
+<td>
+
+```yaml
+# migrator
+__migrator: kind: version
+a:
+  - 1.6
+b:
+  - 1.1
+```
+</td>
+<td> -> </td>
+<td>
+
+```yaml
+# resulting pins
+a:
+  - 1.6
+b:
+  - 1.2
+someflag:
+  - disabled
+```
+</td>
+</tr>
+</table>
+
+These are the most common version migrations and encompass things like moving the pinnings for openssl etc.
+
+Note that the version of `b` was NOT decreased.  
+This is intentional as we don't wish to downgrade accidentally.  
+In practice many migrators could be in flight and may affect the same version.  
+In this case the higher version should always win.  
+If we want to downgrade there is a mechanism but it is purposefully clunky and bad.
+
+#### 2. Matrix Version Addition changes
+
+<table>
+<tr>
+<td>
+
+```yaml
+# Existing pins
+a:
+  - 1.5
+python:
+  - 3.6
+someflag:
+  - disabled
+```
+</td>
+<td> with </td>
+<td>
+
+```yaml
+# migrator
+__migrator: kind: version
+python:
+  - 3.6
+  - 3.8a1
+```
+</td>
+<td> -> </td>
+<td>
+
+```yaml
+# resulting pins
+a:
+  - 1.5
+b:
+  - 3.6
+  - 3.8a1
+someflag:
+  - disabled
+```
+</td>
+</tr>
+</table>
+
+These are things like the compiler migration and can be thought of as additions to a version.  
+These need substantial care when dealing with things like zip_keys.  
```

Comment:
What does this mean?

Reply:
@mariusvniekerk

Reply:
I think the issue here is that if you try to add a pinning under `zip_keys` then you need to make certain that the other pinnings controlled by that zip are similarly expanded otherwise they will be mismatched.

Reply:
The main concern is that it is very easy to accidentally end up with zip_keys for members that don't have the same size.    Whilst it can certainly be done using cfep9 it should be done carefully.